// moonshine-core/src/index.js
var loggerWrapper = require("./loggerwrapper"),
    ComponentLoader = require("./componentLoader"),
    when = require("when"),
    async = require("async"),
    util = require("util")

var componentLoader = new ComponentLoader(loggerWrapper.logger)
var hooks = {}
module.exports.load = function() {
    var apps = Array.prototype.slice.call(arguments,0)
    if (arguments[0] && util.isArray(arguments[0])) {
        apps = arguments[0]
    }
    loggerWrapper.logger.info("loading moonshine")
    var promise = componentLoader.load.apply(componentLoader,apps)

    return promise.then(function(){return;})
}
module.exports.addBaseApp = function(app){
    componentLoader.apps = componentLoader.apps.concat(app)
}
module.exports.runCommand = function(command,apps,args,cb) {
    if (!apps || apps.length <=0) {
        return when.reject(new Error("must provide at least one application"))
    }
    args = args || []
    var deferred  = when.defer()
    module.exports.load(apps).then(function(){
        componentLoader.runCommand(command,args)
            .then(deferred.resolve,deferred.reject)
    }).otherwise(deferred.reject)
    if (cb) deferred.promise.then(cb,cb)
    return deferred.promise;

}
module.exports.settings = componentLoader.settings;

////// service locator
module.exports.service= {}
module.exports.registerService = function(name,service) {
    module.exports.service[name] = service
    module.exports[name] = service
}

module.exports.addHook = function(name,hook) {
    if (!hooks[name]) hooks[name] = []
    hooks[name].push(hook)
}
module.exports.clearHooks = function(){
    hooks = {}
}
module.exports.signal = function(name,args,cb){
    if (!args) args = []
    if (!util.isArray(args)) args = [args]
    var deferred = when.defer()
    async.eachSeries(hooks[name] || [],function(hook,done){
        hook.apply(null,args.concat(done))
    },function(err){
        if (err) {
            deferred.reject(err);
        } else {
            deferred.resolve();
        }
        if (cb) cb(err)
    })
    return deferred.promise
}
