// moonshine-core-api/src/extension.js
var moonshine = require("moonshine-core")
var settings = moonshine.settings
var logger = moonshine.logFactory()
var baucis = require("baucis")
var helpers = require("moonshine-core/helpers")


module.exports.setup = function setupApi(cb){
    moonshine.registerService("api",{
        resources:{},
        resourceOptions:{},
        createResource: function(name,options){
            if (!options && (typeof name == 'string' || name instanceof String)) options = {singular:name}
            var resource = baucis.rest(options);
            moonshine.api.resources[name] = resource
            moonshine.api.resourceOptions[name] = options
            return resource;
        },
        native: baucis
    })
    cb()
}
module.exports.extend = {
    process:helpers.genericLoadModuleFunction("api",logger),
    after: function(done){
        moonshine.signal("api.resources_loaded",[],done)
    }
}

module.exports.wrapup = function addApiToServer(done){
    try {
        moonshine.server.app.use(settings.API_ROOT_PATH,baucis({version:'0.0.1'}))
        moonshine.signal("api.api_registered",[],done)
    } catch (err) {
        cb(err)
    }
}
