// moonshine-core-static/src/extension.js
var moonshine = require("moonshine-core")
var settings = moonshine.settings
var logger = moonshine.logFactory()
var fs = require("fs"),
    path = require("path"),
    async = require("async")

var paths = {}

module.exports.setup = function(cb){
    moonshine.registerService("static",{
         files: paths
    })
    cb()
}

function collectDirPaths(absoluteDirPath,baseUrl,cb) {
    fs.exists(absoluteDirPath,function(exists){
        if (!exists) return cb()
        fs.readdir(absoluteDirPath,function(err,dirContents){
            if (err) cb(err)
            async.forEachSeries(dirContents
                ,function(filename,cb){
                    fs.stat(path.join(absoluteDirPath,filename),function(err,stat){
                        if(stat.isDirectory()) {
                            collectDirPaths(path.join(absoluteDirPath,filename),baseUrl + "/" + filename,cb)
                        } else {
                            paths[baseUrl + "/" + filename] = path.join(absoluteDirPath,filename)
                            cb()
                        }
                    })
                },cb)
        })
    })
}
module.exports.extend = {
    process: function collectStatic(comp,cb){
        try{
            var staticDir = path.join(path.dirname(comp.filename),"static")
            collectDirPaths(staticDir,"",cb)
        } catch(e){
            cb(e)
        }
    },
    after:function setupRouter(cb){
        moonshine.server.app.use(settings.STATIC_ROOTPATH,function(req,res,next){
            var path = req.path.replace(settings.STATIC_ROOTPATH,"")
            if (paths[path]){
                res.sendfile(paths[path])
            }else {
                next()
            }
        })
        moonshine.signal("static.files_collected",[],cb)
    }
}
