// moonshine-core-persistence/src/extension.js
var moonshine = require("moonshine-core")
var settings = moonshine.settings
var logger = moonshine.logFactory()
var mongoose = require("mongoose")
var helpers = require("moonshine-core/helpers")

module.exports.setup = function setupPersistence(cb){
    moonshine.registerService("db",{
        models: {},
        schemas: {},
        getSchema: function(name){
            var schemas = moonshine.db.schemas;
            if (!schemas[name]) {
                schemas[name] = new mongoose.Schema({})
            }
            return schemas[name];
        },
        Schema: mongoose.Schema,
        native: mongoose
    })
    cb()
}

module.exports.extend = {
    process:helpers.genericLoadModuleFunction("model",logger),
    after: function registerSchemas(cb){
        var schemas = moonshine.db.schemas
        try {
            for (var modelName in schemas) {
                var schema = schemas[modelName]
                moonshine.db.models[modelName] = mongoose.model(modelName,schema)
            }
            moonshine.signal("db.models_loaded").then(function(){
                mongoose.connect(settings.DB_CONNECTION,function(err){
                    if (err) return cb(err)
                    moonshine.signal("db.connected",[],cb)
                });

            },cb)
        } catch (err) {
            cb(err)
        }
    }
}
